File: images/runtimes/nodejs/workload.js
const { execSync, exec } = require("child_process");
const process = require("process");
const readline = require("readline");
const fs = require("fs");

// for snapshot
// this approach relies on that we are currently being executed on cpu 0
// and that other cpus writes to the port before us
// since as of now snapshots are created offline, we are fine
const cpu_count = require("os").cpus().length;
for (var i = 0; i < cpu_count; i++) {
    exec(`taskset -c ${i} outl 124 0x3f0`);
}
execSync("taskset -c 0 outl 124 0x3f0");

execSync("mount -r /dev/vdb /srv");

module.paths.push("/srv/node_modules");
const app = require("/srv/workload");

rl = readline.createInterface({
    input: fs.createReadStream('/dev/ttyS1'),
    crlfDelay: Infinity
});
// signal Firerunner that we are ready to receive requests
execSync("outl 126 0x3f0");

rl.on('line', (line) => {
  var hrstart = process.hrtime()
  let req = JSON.parse(line);
  app.handle(req, function(resp) {
    var hrend = process.hrtime(hrstart)
    resp.runtime_sec = hrend[0];
    resp.runtime_ms = hrend[1] / 1000000;
    let respJS = JSON.stringify(resp);
    process.stdout.write(Buffer.from([respJS.length]));
    process.stdout.write(respJS, "utf8");
  });
});

File: images/runtimes/python2/workload.py
#!/usr/bin/env python2

import sys
import os
import imp
import struct
import json
from subprocess import call, Popen
import multiprocessing as mp

# for snapshot
for i in range(1, mp.cpu_count()):
    Popen('taskset -c %d outl 124 0x3f0'%(i), shell=True)
call('taskset -c 0 outl 124 0x3f0', shell=True)

os.system("mount -r /dev/vdb /srv")

with open('/dev/ttyS1', 'r') as tty:
    # signal firerunner we are ready
    call('outl 126 0x3f0', shell=True)

    sys.path.append('/srv/package')
    app = imp.load_source('app', '/srv/workload')

    while True:
        request = json.loads(tty.readline())

        response = app.handle(request)

        responseJson = json.dumps(response)

        sys.stdout.write(struct.pack("@B", len(responseJson)))
        sys.stdout.write(bytes(responseJson))
        sys.stdout.flush()
